src/shapes/src/Quarterpipe.cpp

```diff
@@ -0,0 +1,146 @@
+#include <utils/math/coordinate_transformation.hpp>
+#include <utils/math/vec_rotate.hpp>
+
+namespace Shapes {
+
+void Quarterpipe::calculate_dist(const Utils::Vector3d &pos, double &dist,
+                                 Utils::Vector3d &vec) const {
+
+  // Please have a look at the Doxygen page of this shape for a detailed sketch
+  // and explanation of terms used below
+
+  auto pos_shifted = pos - m_center;
+  auto rel_z = pos_shifted * m_axis;
+
+  // first, do only the 2d problem, add the axis projection later
+
+  auto pos_projected = pos_shifted - m_axis * rel_z;
+  auto pos_norm = pos_projected.norm();
+
+  if (pos_norm == 0.) {
+    dist = m_radius;
+    vec = -dist * m_orientation;
+  } else {
+    // go to cylindrical coordinates (r,delta_phi) where m_axis is the z-axis
+    // and  m_orientation defines the delta_phi = 0 axis
+    auto x = pos_projected * m_orientation;
+    auto y = pos_projected * Utils::vector_product(m_axis, m_orientation);
+    auto pos_delta_phi = std::atan2(y, x);
+
+    // point 1 and 2 are the end points of the shape where the cylinder
+    // intersects the cuboid. Their position vectors also define normals of the
+    // cuboid side planes 1 and 2
+    auto to_point_1 =
+        m_radius * vec_rotate(m_axis, -0.5 * m_delta_phi, m_orientation);
+    auto to_point_2 =
+        m_radius * vec_rotate(m_axis, 0.5 * m_delta_phi, m_orientation);
+
+    if (pos_delta_phi <= -0.5 * m_delta_phi) {
+      // Zone 1
+      vec = pos_projected - to_point_1;
+      dist = vec.norm();
+    } else if (pos_delta_phi >= 0.5 * m_delta_phi) {
+      // Zone 2
+      vec = pos_projected - to_point_2;
+      dist = vec.norm();
+    } else {
+      if (pos_norm < m_radius) {
+        // Zone 3
+        dist = m_radius - pos_norm;
+        vec = -dist * pos_projected / pos_norm;
+      } else {
+        // Zones 4 to 9 are hard to distingish, so all distances are calculated
+        // and the minimum chosen
+
+        // Zone 4, 5, 7 or 9
+        auto dist_plane_1 = (pos_projected * to_point_1 / to_point_1.norm()) -
+                            to_point_1.norm();
+        auto vec_plane_1 = dist_plane_1 * to_point_1 / to_point_1.norm();
+
+        // Zone 4, 6, 8 or 9
+        auto dist_plane_2 = (pos_projected * to_point_2 / to_point_2.norm()) -
+                            to_point_2.norm();
+        auto vec_plane_2 = dist_plane_2 * to_point_2 / to_point_2.norm();
+
+        // Distinguish the Zones
+        if (dist_plane_1 >= 0. and dist_plane_2 >= 0.) {
+          // Zone 9
+          auto corner_point = std::sqrt(2.) * m_radius * m_orientation;
+          vec = pos_projected - corner_point;
+          dist = vec.norm();
+        } else if (dist_plane_1 < 0. and dist_plane_2 < 0.) {
+          // Zone 4, 5 or 6
+
+          // First, find which side is closer
+          // Distinguishes zones 4 and 5 from 4 and 6
+          bool in_zone_4_or_5 =
+              Utils::abs(dist_plane_1) < Utils::abs(dist_plane_2);
```

Comment:
we already know the sign from the `else if` above...
```suggestion
          bool in_zone_4_or_5 = dist_plane_1 > dist_plane_2;
```